docs/language/ql-training-rst/java/query-injection-java.rst

```diff
@@ -0,0 +1,144 @@
+========================
+Example: Query injection
+========================
+
+QL for Java
+
+.. container:: semmle-logo
+
+   Semmle :sup:`TM`
+
+.. rst-class:: setup
+
+Setup
+=====
+
+For this example you should download:
+
+- `QL for Eclipse <https://help.semmle.com/ql-for-eclipse/Content/WebHelp/install-plugin-free.html>`__
+- `VIVO Vitro snapshot <http://downloads.lgtm.com/snapshots/java/vivo-project/Vitro/vivo-project_Vitro_java-srcVersion_47ae42c01954432c3c3b92d5d163551ce367f510-dist_odasa-lgtm-2019-04-23-7ceff95-linux64.zip>`__
+
+.. note::
+
+   For this example, we will be analyzing `VIVO Vitro <https://github.com/vivo-project/Vitro>`__.
+
+   You can also query the project in `the query console <https://lgtm.com/query/project:14040005/lang:java/>`__ on LGTM.com.
+
+   Note that results generated in the query console are likely to differ to those generated in the QL plugin as LGTM.com analyzes the most recent revisions of each project that has been added–the snapshot available to download above is based on an historical version of the code base.
+
+SQL injection
+=============
+
+- Occurs when user input is used to construct an SQL query without any sanitization or escaping.
+
+- Classic example involves constructing a query using string concatenation:
+
+  .. code-block:: sql
+
+     runQuery("SELECT * FROM users WHERE id='" + userId + "'");
+
+
+- If the ``userId`` can be provided by a user, and is not sanitized, then a malicious user can provide input that manipulates the intended query.
+
+- For example, providing the input ``"' OR '1'='1"`` would allow the attacker to return all records in the users table.
+
+.. note::
+
+   `SQL <https://en.wikipedia.org/wiki/SQL>`__ is a database query language, which is often used from within other programming languages to interact with a database. The typical case is that a query is to be executed to find some data, based on some input provided by the user - for example, the user’s ID. However, the interface between the host programming language and SQL is typically implemented by passing a string containing the query to some API.
+
+SPARQL injection
+================
+
+- `SPARQL <https://en.wikipedia.org/wiki/SPARQL>`__ is a language for querying key-value databases in RDF format.
+
+- The same type of vulnerability can occur for SPARQL as for SQL: if the SPARQL query is constructed through string concatenation, a malicious user can subvert the query:
+
+  .. code-block:: sql
+  
+     sparqlAskQuery("ASK { <" + individualURI + "> ?p ?o }");
+
+- SPARQL is used by many projects, but we will be looking at `VIVO Vitro <https://github.com/vivo-project/Vitro/>`__.
+
+.. rst-class:: background2
+
+Developing a QL query
+======================
+
+Finding a query concatenation
```

Comment:
As mentioned elsewhere in this PR, the styling of this slide probably needs a little work still